### PR TITLE
Various improvements.

### PR DESCRIPTION
**Pull request by BillSchumacher**
- Styled and slightly refactored the library
- Added lots of typing
- Fixed logging syntax
- Updated gitignore
- Added option to skip external deps when installing via setup.py
- Updated README
- Added requirements.txt
- Added docstrings to all functions, classes and modules.
- Fixed imports, maybe broke some.

I probably broke some stuff, I ran the test which seems to run but not actually give a good result.

Definitely test this before merging. I got pretty lazy towards the end with typing and such.

More coming soon!

### CODE REVIEW

File: sdkit/train/merge_models.py [hunk 1]
```diff
@@ -1,47 +1,69 @@
...
-           del merged['state_dict']
+            del merged['state_dict']
         save_tensor_file(merged, out_path)
     else:
-        save_tensor_file({'state_dict':merged}, out_path)
+        save_tensor_file({'state_dict': merged}, out_path)
+
 
 # do this pair-wise, to avoid having to load all the models into memory
 def merge_two_models(model0, model1, alpha, use_fp16=True):
```
Comment (thoroc) [inline sdkit/train/merge_models.py:45]: No types on the model?

**Review by thoroc — COMMENTED** [review]
Big bang refactor. It looks ok. But hard to tell with so many files.

File: sdkit/models/model_loader/gfpgan.py [hunk 2]
```diff
@@ -1,16 +1,40 @@
...
+    gfpgan model loader
+"""
 import torch
 from gfpgan import GFPGANer
 
 from sdkit import Context
+from sdkit.utils import log
 
-def load_model(context: Context, **kwargs):
+# hack for a bug in facexlib: https://github.com/xinntao/facexlib/pull/19/files
```
Comment (cmdr2) [inline sdkit/models/model_loader/gfpgan.py:10]: Can you please explain the thinking behind moving this out of `load_model()`?

File: sdkit/models/model_loader/hypernetwork/__init__.py [hunk 3]
```diff
@@ -1,10 +1,19 @@
...
 
 from sdkit import Context
 from sdkit.utils import log, load_tensor_file
+from sdkit.models.model_loader.hypernetwork.hypernetwork import \
+    override_attention_context_kv, HypernetworkModule
 
-def load_model(context: Context, **kwargs):
-    from .hypernetwork import HypernetworkModule, override_attention_context_kv
+
+def load_model(context: Context, **__) -> Union[Dict[str, int], None]:
```
Comment (cmdr2) [inline sdkit/models/model_loader/hypernetwork/__init__.py:13]: TBH, I'm not a fan of the `Union[Dict[str, int], None]` style type declarations.

They're nearly impossible to understand, and the meaning can be conveyed better by a doc-string that says what the return data is.

In this case, it returns the loaded model, so we can just say "returns the loaded model". The exact data type of the model isn't particularly important for this project, and is also not very pythonic.

File: sdkit/models/model_loader/hypernetwork/__init__.py [hunk 4]
```diff
@@ -22,21 +31,24 @@ def load_model(context: Context, **kwargs):
...
+                        model=layers, device=context.device)
+                    for i in range(0, 1)
                 )
 
         override_attention_context_kv(layers)
 
         return layers
     except:
         log.error(traceback.format_exc())
-        log.error(f'Could not load hypernetwork: {model_path}')
```
Comment (cmdr2) [inline sdkit/models/model_loader/hypernetwork/__init__.py:38]: Can you please explain the thinking behind changing the f-strings to %s?

File: sdkit/models/model_loader/hypernetwork/hypernetwork.py [hunk 5]
```diff
@@ -56,7 +95,7 @@ def __init__(self, dim, state_dict=None, layer_structure=None, activation_func=N
 
         self.to(device)
 
-    def fix_old_state_dict(self, state_dict):
+    def fix_old_state_dict(self, state_dict) -> None:
```
Comment (cmdr2) [inline sdkit/models/model_loader/hypernetwork/hypernetwork.py:98]: Again, I'm not a fan of the `-> None` declaration. It means nothing, unless the function really returned something meaningful. And this function intuitively doesn't return anything. The code doesn't need to be more verbose than necessary.

File: sdkit/models/model_loader/realesrgan.py [hunk 6]
```diff
@@ -1,30 +1,49 @@
 import torch
 import os
+
 from basicsr.archs.rrdbnet_arch import RRDBNet
 from realesrgan import RealESRGANer
 
 from sdkit import Context
 
+
```
Comment (cmdr2) [inline sdkit/models/model_loader/realesrgan.py:9]: Why the extra newline?

File: sdkit/models/model_loader/realesrgan.py [hunk 7]
```diff
@@ -1,30 +1,49 @@
...
     return model
 
-def unload_model(context: Context, **kwargs):
-    pass
+
+def unload_model(_: Context, **__):
+    """
+        Unload the model from the context.
+
+        # TODO: Implement this?
```
Comment (cmdr2) [inline sdkit/models/model_loader/realesrgan.py:48]: This stub is only required if the module has some custom tasks to perform, before the cleanup. In this case, realesrgan has nothing extra to do. Hence the `pass` statement.

And these modules (`realesrgan`, `stable-diffusion` etc) are internal modules. They aren't exposed to the user-facing API, so I don't think we need to put redundant documentation in every internal module, when it's clear what they're doing.

Each sub-module inside `model_loader` implements the API of defining: `load_model()` and `unload_model()`. These functions are quite clear from their name, so we don't need redundant documentation for it.

File: sdkit/models/model_loader/stable_diffusion/__init__.py [hunk 8]
```diff
@@ -1,54 +1,81 @@
...
-from torch.nn.functional import silu
 import ldm.modules.attention
 import ldm.modules.diffusionmodules.model
-import tempfile
-from urllib.parse import urlparse
-from pathlib import Path
+from omegaconf import OmegaConf
+from torch.nn.functional import silu
+from transformers import logging as tr_logging
+
```
Comment (cmdr2) [inline sdkit/models/model_loader/stable_diffusion/__init__.py:12]: Why the extra newline?

File: sdkit/models/model_loader/stable_diffusion/__init__.py [hunk 9]
```diff
@@ -1,54 +1,81 @@
...
+from sdkit.utils import load_tensor_file, save_tensor_file, hash_file_quick,\
+     download_file, log
+
+tr_logging.set_verbosity_error()  # suppress unnecessary logging
 
-tr_logging.set_verbosity_error() # suppress unnecessary logging
 
-def load_model(context: Context, scan_model=True, check_for_config_with_same_name=True, **kwargs):
-    from . import optimizations
+def load_model(
```
Comment (cmdr2) [inline sdkit/models/model_loader/stable_diffusion/__init__.py:21]: Not every function signature needs to be multi-line. This one is quite clear and readable in a single line, so we don't have to break it into multiple lines.

File: sdkit/models/model_loader/stable_diffusion/__init__.py [hunk 10]
```diff
@@ -1,54 +1,81 @@
...
+    check_for_config_with_same_name=True,
+    **kwargs
+):
+    """ Load the model. """
+    from sdkit.models.model_loader.stable_diffusion import optimizations
     from sdkit.models import scan_model as scan_model_fn
 
     model_path = context.model_paths.get('stable-diffusion')
-    config_file_path = get_model_config_file(context, check_for_config_with_same_name)
+    config_file_path = get_model_config_file(
```
Comment (cmdr2) [inline sdkit/models/model_loader/stable_diffusion/__init__.py:32]: Same here, why is this on multiple lines?

File: sdkit/models/model_loader/stable_diffusion/__init__.py [hunk 11]
```diff
@@ -1,54 +1,81 @@
...
+        context, check_for_config_with_same_name
+    )
 
     if scan_model:
         scan_result = scan_model_fn(model_path)
         if scan_result.issues_count > 0 or scan_result.infected_files > 0:
-            raise Exception(f'Model scan failed! Potentially infected model: {model_path}')
+            raise RuntimeError(
```
Comment (cmdr2) [inline sdkit/models/model_loader/stable_diffusion/__init__.py:39]: And here

File: sdkit/models/model_loader/stable_diffusion/optimizations.py [hunk 12]
```diff
@@ -93,34 +134,42 @@ def wrap(x):
...
+            return 4  # use for balanced
         elif 'SET_ATTENTION_STEP_TO_6' in context.vram_optimizations:
             return 6
         elif 'SET_ATTENTION_STEP_TO_8' in context.vram_optimizations:
             return 8
         elif 'SET_ATTENTION_STEP_TO_16' in context.vram_optimizations:
             return 16
         elif 'SET_ATTENTION_STEP_TO_24' in context.vram_optimizations:
-            return 24 # use for low
+            return 24  # use for low
```
Comment (cmdr2) [inline sdkit/models/model_loader/stable_diffusion/optimizations.py:172]: Why the extra space?

File: sdkit/models/model_loader/vae.py [hunk 13]
```diff
@@ -1,49 +1,80 @@
...
+def move_model_to_cpu(_: Context):
+    """
+        Move the model to CPU.
 
-def unload_model(context: Context, **kwargs):
+        # TODO: Implement this?
+    """
+
+
+def unload_model(context: Context, **__):
```
Comment (cmdr2) [inline sdkit/models/model_loader/vae.py:56]: Redundant documentation, and in internal modules where the documentation isn't necessary. This bloats up the lines of code unnecessarily.

File: sdkit/models/models_db/__init__.py [hunk 14]
```diff
@@ -1,35 +1,53 @@
...
 
     if db is not None:
         return db
 
     db_path = Path(__file__).parent
     db = {}
-    with open(db_path/'stable_diffusion.json') as f: db['stable-diffusion'] = json.load(f)
-    with open(db_path/'gfpgan.json') as f: db['gfpgan'] = json.load(f)
-    with open(db_path/'realesrgan.json') as f: db['realesrgan'] = json.load(f)
+    with open(db_path/'stable_diffusion.json') as fp:
```
Comment (cmdr2) [inline sdkit/models/models_db/__init__.py:23]: This might be coding-style preference, but I'd like to keep these on a single line for readability of the logic.

File: sdkit/models/models_db/__init__.py [hunk 15]
```diff
@@ -1,35 +1,53 @@
...
+            raise RuntimeError("Could not rebuild index")
         return index.get(quick_hash)
     elif model_id and model_type:
         m = db.get(model_type, {})
         return m.get(model_id)
 
+
 def rebuild_index():
+    """
+        Rebuilds the index of models.
```
Comment (cmdr2) [inline sdkit/models/models_db/__init__.py:48]: Why is this doc-string required? The function name is self-explanatory.

File: sdkit/train/merge_models.py (hunk 1 above)
Comment (cmdr2) [reply sdkit/train/merge_models.py:45]: The doc-string can mention that it's either a Tensor or a string (depicting the path to the model file).

File: sdkit/utils/__init__.py [hunk 16]
```diff
@@ -48,6 +43,45 @@
...
+
+log = logging.getLogger('sdkit')
+LOG_FORMAT = '%(asctime)s.%(msecs)03d %(levelname)s %(threadName)s %(message)s'
+logging.basicConfig(
+        level=logging.INFO,
+        format=LOG_FORMAT,
+        datefmt="%X"
+)
+
+__all__ = [
```
Comment (cmdr2) [inline sdkit/utils/__init__.py:59]: Why is this necessary? The purpose the separate init file in this module is to only import the things that we want to expose. Why do we need a redundant `__all__`?

File: sdkit/utils/hash_utils.py [hunk 17]
```diff
@@ -54,12 +76,16 @@ def compute_quick_hash(total_size_fn, read_bytes_fn):
     Do not use if the file size is less than 3 MB
     '''
     total_size = total_size_fn()
-
-    start_bytes = read_bytes_fn(offset=0x100000, count=0x10000)
-    middle_bytes = read_bytes_fn(offset=int(total_size/2), count=0x10000)
-    end_bytes = read_bytes_fn(offset=total_size - 0x100000, count=0x10000)
+    one_mb = 0x100000
```
Comment (cmdr2) [inline sdkit/utils/hash_utils.py:79]: Nice

File: sdkit/utils/latent_utils.py [hunk 18]
```diff
@@ -1,59 +1,87 @@
...
+    """ Convert PIL image to torch tensor. """
     if img is None:
         return None
 
-    img = np.array(img).astype(np.float32) / 255.0
-    img = img[None].transpose(0, 3, 1, 2)
-    img = torch.from_numpy(img)
-    img = 2. * img - 1. if shift_range else img
-    img = img.to(device)
+    _img: np.ndarray = np.array(img).astype(np.float32) / 255.0
```
Comment (cmdr2) [inline sdkit/utils/latent_utils.py:24]: Why are we using `_img` here?

File: sdkit/utils/memory_utils.py [hunk 19]
```diff
@@ -9,94 +12,141 @@
...
+        psutil.virtual_memory().total
+    vram_free, vram_total = \
+        torch.cuda.mem_get_info(device) if device != 'cpu' else (0, 0)
     vram_used = (vram_total - vram_free)
-
-    ram_used /= 1024**3
-    ram_total /= 1024**3
-    vram_used /= 1024**3
-    vram_total /= 1024**3
+    ram = 1024**3
```
Comment (cmdr2) [inline sdkit/utils/memory_utils.py:38]: Wouldn't `one_gb` be more appropriate, for this variable name?

File: sdkit/utils/memory_utils.py [hunk 20]
```diff
@@ -9,94 +12,141 @@
...
+                (
+                    hasattr(obj, 'data') and
+                    torch.is_tensor(obj.data)
+                )
+            ) and obj.device == device:
                 tensors.append(obj)
-        except:
-            pass
+        except Exception as ex:
+            log.warning('Error while getting tensors in memory: %s', ex)
```
Comment (cmdr2) [inline sdkit/utils/memory_utils.py:135]: This `pass` statement is intentional. Lots of objects in the memory will throw errors, so this function will be unusable otherwise. https://discuss.pytorch.org/t/how-to-debug-causes-of-gpu-memory-leaks/6741/3

File: sdkit/utils/memory_utils.py [hunk 21]
```diff
@@ -106,64 +156,87 @@ def print_largest_tensors_in_memory(device, num=10):
     print(f'== {num} largest tensors on {device} ==')
     print(_fmt_tensors_summary(entries))
     print('---')
-    print(f'Total memory occupied on {device} by {n} tensors: {total_mem:.1f} MiB')
+    print(f'Total memory occupied on {device} by'
```
Comment (cmdr2) [inline sdkit/utils/memory_utils.py:159]: Why the break of lines? We don't need to enforce 80-char line width, most monitors and IDEs easily show wider lines, and we don't need to make the codebase look artificially verbose.

File: sdkit/utils/memory_utils.py [hunk 22]
```diff
@@ -106,64 +156,87 @@ def print_largest_tensors_in_memory(device, num=10):
...
     # print the diff from the prev snapshot
     if is_first_snapshot or not print_snapshot:
         return
 
-    new_tensor_entries = [entry for entry in entries if entry[0] in new_tensor_ids]
-    new_tensors_total_mem = reduce(lambda sum, entry: sum + entry[1], new_tensor_entries, 0) # MiB
+    new_tensor_entries = [
+        entry for entry in entries if entry[0] in new_tensor_ids
+    ]
+    new_tensors_total_mem = reduce(
```
Comment (cmdr2) [inline sdkit/utils/memory_utils.py:192]: Can remain a single line

File: setup.py [hunk 23]
```diff
@@ -1,14 +1,25 @@
...
 import setuptools
 
-setuptools.setup(
-    install_requires=[
-        "stable-diffusion-sdkit", # wrapper around stable-diffusion, to allow pip install
+install_requires = []
+
+
+if not os.getenv('SKIP_EXTERNAL_DEPS'):
+    install_requires: list[str] = [
```
Comment (cmdr2) [inline setup.py:11]: The data-type declaration here is redundant. It's clear that it's a list of strings.

File: setup.py [hunk 24]
```diff
@@ -1,14 +1,25 @@
...
+import os
 import setuptools
 
-setuptools.setup(
-    install_requires=[
-        "stable-diffusion-sdkit", # wrapper around stable-diffusion, to allow pip install
+install_requires = []
+
+
+if not os.getenv('SKIP_EXTERNAL_DEPS'):
```
Comment (cmdr2) [inline setup.py:10]: Can you please explain why we need to check this variable? And what's the use-case for installing sdkit without the dependencies?

File: sdkit/models/model_loader/hypernetwork/__init__.py (hunk 4 above)
Comment (thoroc) [reply sdkit/models/model_loader/hypernetwork/__init__.py:38]: I think that's the reason: https://docs.python.org/3/howto/logging.html#optimization. At least that's why I do too.

Comment (cmdr2) [reply sdkit/models/model_loader/hypernetwork/__init__.py:38]: If we're using expensive function calls, then it makes sense. Otherwise f-strings are preferable since they're clearer to read

File: sdkit/models/model_loader/vae.py (hunk 13 above)
Comment (thoroc) [reply sdkit/models/model_loader/vae.py:56]: Does documentation on method has an impact on how the code run? As it stand it isn't very useful but maybe that need to change to be more descriptive. It will help anyone looking at the code for the first time.

File: sdkit/models/models_db/__init__.py (hunk 14 above)
Comment (thoroc) [reply sdkit/models/models_db/__init__.py:23]: having `with` on the same line as the code run within the context is not usual tbh.

File: sdkit/utils/latent_utils.py (hunk 18 above)
Comment (thoroc) [reply sdkit/utils/latent_utils.py:24]: The local var was reused. Not sure why this was chosen.
